File: app/src/main/java/com/example/ideal/myapplication/adapters/MessageAdapter.java
package com.example.ideal.myapplication.adapters;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.TextView;

import com.example.ideal.myapplication.R;
import com.example.ideal.myapplication.fragments.chatElements.MessageOrderElement;
import com.example.ideal.myapplication.fragments.objects.Message;
import com.example.ideal.myapplication.helpApi.WorkWithTimeApi;
import com.example.ideal.myapplication.other.DBHelper;

import java.util.ArrayList;

public class MessageAdapter extends RecyclerView.Adapter<MessageAdapter.MessageViewHolder> {

    private int numberItems;
    private  ArrayList<Message> messageList;
    private Context context;
    private DBHelper dbHelper;

    public MessageAdapter(int numberItems, ArrayList<Message> messages, DBHelper dbHelper) {
        this.numberItems = numberItems;
        messageList = messages;
        this.dbHelper = dbHelper;
    }

    @Override
    public MessageViewHolder onCreateViewHolder(ViewGroup viewGroup, int i) {
        // в этом методе создаем ViewHolder
        context = viewGroup.getContext();
        int layoutIdForListItem = R.layout.message_element;
        //Класс, который позволяет создавать представления из xml файла
        LayoutInflater layoutInflater = LayoutInflater.from(context);
        // откуда, куда, необходимо ли помещать в родителя
        View view = layoutInflater.inflate(layoutIdForListItem,viewGroup,false);

        MessageViewHolder viewHolder = new MessageViewHolder(view);
        return viewHolder;
    }

    @Override
    public void onBindViewHolder(@NonNull MessageViewHolder messageViewHolder, int index) {
        messageViewHolder.bind(messageList.get(index));
    }

    @Override
    public int getItemCount() {
        //общее количество элементов в нашем списке
        return numberItems;
    }

    class MessageViewHolder extends RecyclerView.ViewHolder {

        View view;
        public MessageViewHolder(@NonNull View itemView) {
            super(itemView);
            view = itemView;
        }
        void bind(Message message){
            MessageOrderElement messageOrderElement = new MessageOrderElement(message,view,context,dbHelper);

        }
    }

}

File: app/src/main/java/com/bunbeauty/ideal/myapplication/clean_architecture/mvp/adapters/MessageAdapter.kt
package com.bunbeauty.ideal.myapplication.clean_architecture.mvp.adapters

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.RecyclerView.ViewHolder
import com.android.ideal.myapplication.R
import com.bunbeauty.ideal.myapplication.clean_architecture.mvp.adapters.MessageAdapter.MessageViewHolder
import com.bunbeauty.ideal.myapplication.clean_architecture.mvp.adapters.elements.chatElements.messageElements.MessageCancelElement
import com.bunbeauty.ideal.myapplication.clean_architecture.mvp.adapters.elements.chatElements.messageElements.MessageServiceReviewElement
import com.bunbeauty.ideal.myapplication.clean_architecture.mvp.adapters.elements.chatElements.messageElements.MessageTextElement
import com.bunbeauty.ideal.myapplication.clean_architecture.mvp.adapters.elements.chatElements.messageElements.MessageUserReviewElement
import com.bunbeauty.ideal.myapplication.clean_architecture.data.db.models.entity.Message
import com.bunbeauty.ideal.myapplication.clean_architecture.mvp.presenters.chat.MessagesPresenter

class MessageAdapter : RecyclerView.Adapter<MessageViewHolder>() {

    private var messageList = mutableListOf<Message>()
    private lateinit var messagesPresenter: MessagesPresenter

    fun setData(messagesPresenter: MessagesPresenter) {
        this.messagesPresenter = messagesPresenter
    }

    fun addItemToStart(message: Message) {
        messageList.add(message)
        messageList.sortBy { it.time }
        notifyItemInserted(0)
    }

    fun addItemToBottom(message: Message) {
        messageList.add(message)
        notifyItemInserted(messageList.size)
    }

    fun updateMessageAdapter(message: Message) {
        val foundMessage = messageList.find { it.id == message.id }
        if (foundMessage != null) {
            val index = messageList.indexOf(foundMessage)
            messageList[index] = message
            notifyItemChanged(index)
        }
    }

    fun removeMessageAdapter(message: Message) {
        val foundMessage = messageList.find { it.id == message.id }
        if (foundMessage != null) {
            val index = messageList.indexOf(foundMessage)
            messageList.remove(foundMessage)
            notifyItemRemoved(index)
        }
    }

    override fun onCreateViewHolder(viewGroup: ViewGroup, i: Int): MessageViewHolder {
        val context = viewGroup.context
        val layoutIdForListItem = R.layout.element_message
        val layoutInflater = LayoutInflater.from(context)
        val view = layoutInflater.inflate(layoutIdForListItem, viewGroup, false)
        return MessageViewHolder(view, context)
    }

    override fun onBindViewHolder(
        messageViewHolder: MessageViewHolder,
        index: Int
    ) {
        messageViewHolder.bind(messageList[index])
    }

    override fun getItemCount() = messageList.size

    inner class MessageViewHolder(private val view: View, private val context: Context) :
        ViewHolder(view) {

        fun bind(message: Message) {

            when (message.type) {
                Message.TEXT_STATUS -> {
                    val messageTextElement = MessageTextElement(message, view)
                    messageTextElement.setIsMyMessage(message, view)
                    messageTextElement.setButtonVisibility(message, view, "")
                }

                Message.SERVICE_REVIEW_STATUS -> {
                    val messageTextElement =
                        MessageServiceReviewElement(messagesPresenter, message, view)
                    messageTextElement.setIsMyMessage(message, view)
                    messageTextElement.setButtonVisibility(
                        message,
                        view,
                        context.getString(R.string.rate)
                    )
                }

                Message.USER_REVIEW_STATUS -> {
                    val messageUserReviewElement =
                        MessageUserReviewElement(messagesPresenter, message, view)
                    messageUserReviewElement.setIsMyMessage(message, view)
                    messageUserReviewElement.setButtonVisibility(
                        message,
                        view,
                        context.getString(R.string.rate)
                    )
                }

                Message.CANCEL_STATUS -> {
                    val messageCancelElement =
                        MessageCancelElement(messagesPresenter, message, view)
                    messageCancelElement.setIsMyMessage(message, view)
                    messageCancelElement.setButtonVisibility(
                        message,
                        view,
                        context.getString(R.string.cancel)
                    )
                }

                else -> {
                    val messageTextElement = MessageTextElement(message, view)
                    messageTextElement.setIsMyMessage(message, view)
                    messageTextElement.setButtonVisibility(message, view, "")
                }
            }
        }
    }

}